### ISSUE TITLE
Problem with TestChannelBinder when mixing `@EnableBinding` and function.

### ISSUE THREAD

**Issue by dturanski** [spring-cloud/spring-cloud-stream#1876]
Sending two messages fails when using `@EnableBinding` with registered function. First message works, the second appears to bypass the function altogether.  See https://github.com/dturanski/two-message-bug-sample

**Comment by olegz** [comment]
Finally got a chance to look. . .

What you're trying to do is not supported and that's by design, but you can accomplish it with spring profiles. Basically you're trying to bootstrap _this or that_ configuration based on the property value while we are looking at the presence of EnableBinding annotation to determine whether to bootstrap or not to bootstrap functions  (well it's not about "not to bootstrap", rather how. . .).

***Functional OR Annotation*** is generally a mutually exclusive proposition primarily due to how the bindings are created. In annotation you explicitly name them while in function they are derived based on the function name. There is more to it but i'll stop now otherwise it's a very long explanation ;)

The main point though is that the support for ***Function WITH Annotation*** was specifically added to support function composition from the existing annotation-based elements. In other words to support syntax like `spring.cloud.function.definition=|myFunction` which presumes there is an existing  `Source` or `spring.cloud.function.definition=myFunction|` which presumes there is an existing `Sink`.

While it may be possible to support both side by side it seems to me like a step backward. Also, as we did attempt to support it initially it did result in very hard to follow and maintain code.

**Comment by dturanski** [comment]
@olegz - Does this mean that the original scenario using `spring.cloud.stream.function.definition` with an existing `@EnableBinding` Source is no longer supported?

**Comment by olegz** [comment]
@dturanski it is supported. You can have a source and have a function definition as `spring.cloud.stream.function.definition=|myFunction`. Note the pipe in front of `myFunction`.

**Comment by mgalindor** [comment]
@olegz  would be fine if you add the behavior in the documentation

A note like this  `Functional OR Annotation is generally a mutually exclusive` can save us a lot of time .

Regards
